elkscmd/man/man1/meminfo.1
.TH meminfo 1
.SH NAME
meminfo \- list system memory usage
.SH SYNOPSIS
.B meminfo
.RB [ \-a ]
.RB [ \-m ]
.RB [ \-f ]
.RB [ \-t ]
.RB [ \-b ]
.RB [ \-s ]
.RB [ \-h ]
.br
.SS OPTIONS
Defaults to showing all memory (equivalent to -aftbs).
.TP 5
.B -a
Show memory in use by applications
.TP 5
.B -m
Show main memory sorted by segment
.TP 5
.B -f
Show free memory
.TP 5
.B -t
Show tty and driver memory
.TP 5
.B -b
Show system buffer and pipe memory
.TP 5
.B -s
Show system task, inode and file memory
.TP 5
.B -h
Show help
.SH DESCRIPTION
.B meminfo
traverses the kernel local heap and displays a line for each in-use or free entry. 
Since the local heap also manages the system memory outside of the kernel
normally available for running application programs, the in-use or free
segments of system memory outside the kernel
(sometimes called external or main memory) will also be displayed.
.PP
For every line, the local heap address, type and size are displayed.
For external memory, the external memory segment, type, size and access count
are also displayed. Each of these are detailed below.
.PP
At the bottom of the listing, the total size and free size of the kernel
local heap are displayed, along with the external (main) memory system total,
used and free space in kilobytes (KB).
.PP
By inspecting the external (main) memory entries, one can determine
how much contiguous memory is available for running additional programs,
as well as the amount of memory used for kernel system buffers, and
application program code and data segments.
.PP
Each line of the listing contains:
.TP 10
HEAP
The kernel local heap address.
.TP 10
TYPE
The type of memory allocated (or free).
.TP 10
SIZE
The size in bytes of the allocated or free local memory.
.PP
For external/main memory (type SEG), the following additional items are displayed:
.TP 10
SEG
The segment address of the external/main memory.
.TP 10
TYPE
What the external/main memory segment is used for (or free).
.TP 10
SIZE
The size in bytes of the allocated or free external/main memory segment.
.TP 10
CNT
The access count (0 is free, >1 is shared).
.SH "LOCAL HEAP TYPES"
Local memory allocation types can be one of the following:
.TP 10
SEG
External (main) memory segment.
.TP 10
TTY
TTY input or output buffer.
.TP 10
DRVR
Direct Floppy or other driver buffers.
.TP 10
BUFH
System buffer headers. The buffer data contents may be in main or XMS memory.
.TP 10
PIPE
Open pipe buffer.
.TP 10
TASK
Kernel per process data structure.
.TP 10
INOD
System inodes.
.TP 10
FILE
System open files.
.TP 10
free
Unallocated memory in the local heap.
.SH "EXTERNAL (MAIN) MEMORY TYPES"
Main memory allocation types can be one of the following:
.TP 10
CSEG
Running application code segment.
.TP 10
DSEG
Running application data segment.
.TP 10
BUF
Block subsystem buffer data, 1024 bytes per buffer.
.TP 10
RDSK
Ramdisk data.
.TP 10
free
Unallocated main memory.
.SH FILES
.TP 10
.B /dev/kmem
.SH BUGS
.B meminfo
does not yet display XMS memory usage.
.SH AUTHOR
Greg Haerr (greg@censoft.com)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sys_utils/meminfo.c
/*
 * meminfo.c
 *
 * Copyright (c) 2002 Harry Kalogirou
 * harkal@gmx.net
 *
 * Enhanced by Greg Haerr 24 Apr 2020
 *
 * This file may be distributed under the terms of the GNU General Public
 * License v2, or at your option any later version.
 */
#define __LIBC__            /* get all typedefs */
#include <linuxmt/types.h>
#include <linuxmt/mm.h>
#include <linuxmt/mem.h>
#include <linuxmt/heap.h>
#include <linuxmt/sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>

#define LINEARADDRESS(off, seg)     ((off_t) (((off_t)seg << 4) + off))

int aflag;      /* show application memory*/
int fflag;      /* show free memory*/
int tflag;      /* show tty and driver memory*/
int bflag;      /* show buffer memory*/
int sflag;      /* show system memory*/
int mflag;      /* show main memory*/
int allflag;    /* show all memory*/

unsigned int ds;
unsigned int heap_all;
unsigned int seg_all;
unsigned int taskoff;
int maxtasks;
struct task_struct task_table;

int memread(int fd, word_t off, word_t seg, void *buf, int size)
{
    if (lseek(fd, LINEARADDRESS(off, seg), SEEK_SET) == -1)
        return 0;

    if (read(fd, buf, size) != size)
        return 0;

    return 1;
}

word_t getword(int fd, word_t off, word_t seg)
{
    word_t word;

    if (!memread(fd, off, seg, &word, sizeof(word)))
        return 0;
    return word;
}

void process_name(int fd, unsigned int off, unsigned int seg)
{
    word_t argc, argv;
    char buf[80];

    argc = getword(fd, off, seg);

    while (argc-- > 0) {
        off += 2;
        argv = getword(fd, off, seg);
        if (!memread(fd, argv, seg, buf, sizeof(buf)))
            return;
        printf("%s ",buf);
        break;      /* display only executable name for now */
    }
}

struct task_struct *find_process(int fd, unsigned int seg)
{
    int i;
    int off = taskoff;

    for (i = 0; i < maxtasks; i++) {
        if (!memread(fd, off, ds, &task_table, sizeof(task_table))) {
            perror("taskinfo");
            exit(1);
        }
        if ((unsigned)task_table.mm[SEG_CODE] == seg ||
            (unsigned)task_table.mm[SEG_DATA] == seg) {
            return &task_table;
        }
        off += sizeof(struct task_struct);
    }
    return NULL;
}

static long total_segsize = 0;
static char *segtype[] =
    { "free", "CSEG", "DSEG", "DDAT", "FDAT", "BUF ", "RDSK" };

void display_seg(int fd, word_t mem)
{
    seg_t segbase = getword(fd, mem + offsetof(segment_s, base), ds);
    segext_t segsize = getword(fd, mem + offsetof(segment_s, size), ds);
    word_t segflags = getword(fd, mem + offsetof(segment_s, flags), ds) & SEG_FLAG_TYPE;
    byte_t ref_count = getword(fd, mem + offsetof(segment_s, ref_count), ds);
    struct task_struct *t;

    printf("   %4x   %s %7ld %4d  ",
        segbase, segtype[segflags], (long)segsize << 4, ref_count);
    if (segflags == SEG_FLAG_CSEG || segflags == SEG_FLAG_DSEG) {
        if ((t = find_process(fd, mem)) != NULL) {
            process_name(fd, t->t_begstack, t->t_regs.ss);
        }
    }

    total_segsize += (long)segsize << 4;
}

void dump_segs(int fd)
{
    word_t n, mem;
    seg_t segbase, oldbase = 0;

    printf("    SEG   TYPE    SIZE  CNT  NAME\n");
    n = getword (fd, seg_all + offsetof(list_s, next), ds);
    while (n != seg_all) {
        mem = n - offsetof(segment_s, all);
        segbase = getword(fd, mem + offsetof(segment_s, base), ds);
        if (segbase < oldbase) printf("\n");
        oldbase = segbase;
        display_seg(fd, mem);
        printf("\n");

        /* next in list */
        n = getword(fd, n + offsetof(list_s, next), ds);
    }
}

void dump_heap(int fd)
{
    word_t total_size = 0;
    word_t total_free = 0;
    static char *heaptype[] =
        { "free", "MEM ", "DRVR", "TTY ", "TASK", "BUFH", "PIPE", "INOD", "FILE", "CACH"};

    printf("  HEAP   TYPE  SIZE    SEG   TYPE    SIZE  CNT  NAME\n");

    word_t n = getword (fd, heap_all + offsetof(list_s, next), ds);
    while (n != heap_all) {
        word_t h = n - offsetof(heap_s, all);
        word_t size = getword(fd, h + offsetof(heap_s, size), ds);
        byte_t tag = getword(fd, h + offsetof(heap_s, tag), ds) & HEAP_TAG_TYPE;
        word_t mem = h + sizeof(heap_s);
        word_t segflags;
        int free, app, tty, buffer, system;

        if (tag == HEAP_TAG_SEG)
            segflags = getword(fd, mem + offsetof(segment_s, flags), ds) & SEG_FLAG_TYPE;
        else segflags = -1;
        free = (tag == HEAP_TAG_FREE || segflags == SEG_FLAG_FREE);
        app = ((tag == HEAP_TAG_SEG)
            && (segflags == SEG_FLAG_CSEG || segflags == SEG_FLAG_DSEG ||
                segflags == SEG_FLAG_DDAT || segflags == SEG_FLAG_FDAT));
        tty = (tag == HEAP_TAG_TTY || tag == HEAP_TAG_DRVR);
        buffer = (tag == HEAP_TAG_SEG && segflags == SEG_FLAG_EXTBUF)
            || tag == HEAP_TAG_BUFHEAD || tag == HEAP_TAG_CACHE || tag == HEAP_TAG_PIPE;
        system = (tag == HEAP_TAG_TASK || tag == HEAP_TAG_INODE || tag == HEAP_TAG_FILE);

        if (allflag || (fflag && free) || (aflag && app) || (tflag && tty)
                    || (bflag && buffer) || (sflag && system)) {
            printf("  %4x   %s %5d", mem, heaptype[tag], size);
            total_size += size + sizeof(heap_s);
            if (tag == HEAP_TAG_FREE)
                total_free += size;

            switch (tag) {
            case HEAP_TAG_SEG:
                display_seg(fd, mem);
                break;
            }
            printf("\n");
        }

        /* next in heap*/
        n = getword(fd, n + offsetof(list_s, next), ds);
    }

    printf("  Heap/free   %5u/%5u Total mem %7ld\n", total_size, total_free, total_segsize);
}

void usage(void)
{
    printf("usage: meminfo [-amftbsh]\n");
}

int main(int argc, char **argv)
{
    int fd, c;
    struct mem_usage mu;

    if (argc < 2)
        allflag = 1;
    else while ((c = getopt(argc, argv, "amftbsh")) != -1) {
        switch (c) {
            case 'a':
                aflag = 1;
                break;
            case 'm':
                mflag = 1;
                break;
            case 'f':
                fflag = 1;
                break;
            case 't':
                tflag = 1;
                break;
            case 'b':
                bflag = 1;
                break;
            case 's':
                sflag = 1;
                break;
            case 'h':
                usage();
                return 0;
            default:
                usage();
                return 1;
        }
    }

    if ((fd = open("/dev/kmem", O_RDONLY)) < 0) {
        perror("meminfo");
        return 1;
    }
    if (ioctl(fd, MEM_GETDS, &ds) ||
        ioctl(fd, MEM_GETHEAP, &heap_all) ||
        ioctl(fd, MEM_GETSEGALL, &seg_all) ||
        ioctl(fd, MEM_GETTASK, &taskoff) ||
        ioctl(fd, MEM_GETMAXTASKS, &maxtasks)) {
          perror("meminfo");
          return 1;
    }
    if (!memread(fd, taskoff, ds, &task_table, sizeof(task_table))) {
        perror("taskinfo");
    }
    if (mflag)
        dump_segs(fd);
    else dump_heap(fd);

    if (!ioctl(fd, MEM_GETUSAGE, &mu)) {
        /* note MEM_GETUSAGE amounts are floors, so total may display less by 1k than actual*/
        printf("  Memory usage %4dKB total, %4dKB used, %4dKB free\n",
            mu.used_memory + mu.free_memory, mu.used_memory, mu.free_memory);
    }

    return 0;
}
